// swarm-decentralized-chat/dist/asyncQueueOld.js
import { FIRST_SEGMENT_INDEX } from "./constants";
import { incrementHexString, sleep } from "./utils";
export class AsyncQueue {
    constructor(settings = {}) {
        this.isProcessing = false;
        this.currentPromiseProcessing = false;
        this.isWaiting = false;
        this.queue = [];
        this.indexed = settings.indexed || false;
        this.index = settings.index || FIRST_SEGMENT_INDEX;
        this.waitable = settings.waitable || false;
        this.clearWaitTime = settings.clearWaitTime || 100;
    }
    async processQueue() {
        if (this.isProcessing)
            return;
        this.isProcessing = true;
        while (this.queue.length > 0) {
            this.currentPromiseProcessing = true;
            const promise = this.queue.shift();
            const action = this.indexed ? () => promise(this.index) : () => promise();
            if (this.waitable) {
                try {
                    await action();
                    this.index = incrementHexString(this.index);
                }
                catch (error) {
                    console.error('Error processing promise:', error);
                }
                finally {
                    this.currentPromiseProcessing = false;
                }
            }
            else {
                action()
                    .then(() => {
                    this.index = incrementHexString(this.index);
                })
                    .catch((error) => {
                    console.error('Error processing promise:', error);
                })
                    .finally(() => {
                    this.currentPromiseProcessing = false;
                });
            }
        }
        this.isProcessing = false;
    }
    enqueue(promiseFunction) {
        this.queue.push(promiseFunction);
        this.processQueue();
    }
    async clearQueue() {
        this.queue = [];
        while (this.isProcessing || this.currentPromiseProcessing) {
            await sleep(this.clearWaitTime);
        }
    }
    async waitForProcessing() {
        if (this.isWaiting)
            return true;
        this.isWaiting = true;
        while (this.isProcessing || this.currentPromiseProcessing) {
            await sleep(this.clearWaitTime);
        }
        this.isWaiting = false;
    }
}
